File: src/Python/disruptor/engine.py
from __future__ import annotations
import threading
import queue
from typing import Callable, TypeVar, Generic, Optional

T = TypeVar('T')

class DisruptorEngine(Generic[T]):
    def __init__(self, name: str, pipeline: Callable[[T], T], capacity: int = 8192):
        self.name = name
        self.pipeline = pipeline
        self.q: queue.Queue[T] = queue.Queue(maxsize=capacity)
        self._running = threading.Event()
        self._running.set()
        self._worker = threading.Thread(target=self._run, name=f"{name}-worker", daemon=True)
        self._worker.start()

    def _run(self):
        while self._running.is_set():
            try:
                item = self.q.get(timeout=0.1)
            except queue.Empty:
                continue
            try:
                self.pipeline(item)
            finally:
                self.q.task_done()

    def publish(self, payload: T) -> None:
        if not self._running.is_set():
            raise RuntimeError("engine stopped")
        # Backpressure policy: block until space is available
        self.q.put(payload, block=True)

    def shutdown(self) -> None:
        self._running.clear()

    def close(self) -> None:
        self.shutdown()

File: src/typescript/disruptor/engine.ts
export class DisruptorEngine<T> {
  private name: string;
  private pipelineCallable: (payload: T) => unknown | Promise<unknown>;
  private capacity: number;
  private queue: Array<T>;
  private running: boolean;
  private pendingPublishes: Array<{ payload: T; resolve: () => void }>;

  constructor(name: string, pipelineCallable: (payload: T) => unknown | Promise<unknown>, capacity: number = 8192) {
    this.name = name;
    this.pipelineCallable = pipelineCallable;
    this.capacity = capacity;
    this.queue = [];
    this.running = true;
    this.pendingPublishes = [];
    this.runLoop();
  }

  private async runLoop(): Promise<void> {
    // Single-consumer loop
    while (this.running) {
      const nextItem = this.queue.shift();
      if (nextItem === undefined) {
        await new Promise<void>(resolve => setTimeout(resolve, 10));
        continue;
      }
      try {
        await this.pipelineCallable(nextItem);
      } finally {
        this.drainPendingPublishes();
      }
    }
  }

  private drainPendingPublishes(): void {
    while (this.queue.length < this.capacity && this.pendingPublishes.length > 0) {
      const next = this.pendingPublishes.shift();
      if (!next) {
        break;
      }
      this.queue.push(next.payload);
      next.resolve();
    }
  }

  async publish(payload: T): Promise<void> {
    if (!this.running) {
      throw new Error("engine stopped");
    }
    if (this.queue.length < this.capacity) {
      this.queue.push(payload);
      return;
    }
    // Backpressure: block (via promise) until space is available
    await new Promise<void>(resolve => {
      this.pendingPublishes.push({ payload, resolve });
    });
  }

  shutdown(): void {
    this.running = false;
  }

  close(): void {
    this.shutdown();
  }
}
